### PR TITLE
Refactor utils.py

### LINKED ISSUE #490: Project hierarchy is prone to circular dependencies
Circular dependencies occur when `module_a` depends on `module_b` and `module_b` also depends on `module_a` (https://stackabuse.com/python-circular-imports/).

While looking at #484 I noticed that `__ivadomed_dir__` is defined in multiple places in the project and proposed to clean it up, but while doing so I uncovered some problems especially in `utils.py`:
- `utils.py` is doing too much. Classic problem of this type of module (see https://github.com/neuropoly/spinalcordtoolbox/issues/2898).
- `utils.py` is importing 6 other `ivadomed` modules. Usually `utils.py` is something that provides non-specific/general helper functionality. It is the kind of module that is at the bottom of the import hierarchy. By depending on these 6 modules we now need to be careful that they do not depend on `utils.py` in the future, otherwise it may cause circular dependencies (e.g https://github.com/ivadomed/ivadomed/pull/484/checks?check_run_id=1309819496)

To avoid these types of problem it is good to have a proper dependency hierarchy: i.e `a` depends on `b` and `b` depends on `c`, everything can depend on `c`, but `c` cannot depend on `a` or `b`...

It is probably a good idea to:
- refactor module-specific functions in the appropriate module (e.g `get_metric_fns()`, why is that in utils and not `metrics.py`?)
- evaluate if what is remaining makes sense to exist in `utils.py` or if a dedicated module should be added (`image.py` maybe?)
- further split `utils.py` if necessary (although I doubt it would be the case here).

### PR DESCRIPTION
**Pull request by lrouhier**
utils.py file started to get really big and lots of functions could have been in separate files. This PR aims at creating new files to refactor utils.py. This will break retro compatibility.

This PR create the following files:

* uncertainty.py: run_uncertainty, combine_predictions, voxelwise_uncertainty, structurewise_uncertainty
* mixup.py : mixup, save_mixup_sample
* inference.py: pred_to_nib, segment_volume, volume_reconstruction, onnx_inference
* visualization.py: animated_gif, looping_Pillow_writer, overlap_im_seg, save_color_label, convert_labels_to_RGB, save_tensorboard_img, save_feature_map, HookBasedFeatureExtractor

Move function:
* Get_metric_fns to metrics.py
 
It also changes the import and function call in files when needed. 
Documentation has been modified accordingly. Unneeded imports in documentation were also replaced. 
 

fix #490

### CODE REVIEW

File: ivadomed/utils.py [hunk 1]
```diff
@@ -714,141 +140,19 @@ def save_feature_map(batch, layer_name, log_directory, model, test_input, slice_
...
-        writer.add_image(dataset_type + '/Predictions', grid_img, epoch)
-
-        grid_img = vutils.make_grid(convert_labels_to_RGB(gt_samples),
-                                    normalize=True,
-                                    scale_each=True)
-
-        writer.add_image(dataset_type + '/Ground Truth', grid_img, epoch)
-
-
 class SliceFilter(object):
```
Comment (andreanne-lemay) [inline ivadomed/utils.py:156]: Move to loader/utils?

**Review by andreanne-lemay — COMMENTED** [review]
Great thank you so much! Looks good in general, I did a few suggestions. Also, files `/dev` should be updated with the new locations of functions

File: ivadomed/image.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import nibabel as nib
```
[the remark this answers is not in the document]
Comment (lrouhier) [reply ivadomed/image.py:1]: I don't see it so much as a problem. we have other file that are really small, and it is more for "future development" and future functions.

File: ivadomed/utils.py [hunk 3]
```diff
@@ -657,29 +106,6 @@ def forward(self, x):
...
-    # https://gitship.com/neuroscience/nibabel/blob/master/nibabel/orientations.py
-    ref_orientation = nib.orientations.io_orientation(nib_ref.affine)
-    ras_orientation = nib.orientations.io_orientation(nib_ref_canonical.affine)
-    # Return the orientation that transforms from ras to ref_orientation
-    trans_orient = nib.orientations.ornt_transform(ras_orientation, ref_orientation)
-    # apply transformation
-    return nib.orientations.apply_orientation(arr_ras, trans_orient)
-
-
 def save_feature_map(batch, layer_name, log_directory, model, test_input, slice_axis):
```
[the remark this answers is not in the document]
Comment (lrouhier) [reply ivadomed/utils.py]: done in  94816c8 and  9d89939

File: ivadomed/utils.py (hunk 1 above)
Comment (lrouhier) [reply ivadomed/utils.py]: done in 772d87f

File: ivadomed/image.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    # Orient image in RAS according to slice axis
+    arr_ras = imed_loader_utils.orient_img_ras(arr, slice_axis)
+
+    # https://gitship.com/neuroscience/nibabel/blob/master/nibabel/orientations.py
+    ref_orientation = nib.orientations.io_orientation(nib_ref.affine)
+    ras_orientation = nib.orientations.io_orientation(nib_ref_canonical.affine)
+    # Return the orientation that transforms from ras to ref_orientation
+    trans_orient = nib.orientations.ornt_transform(ras_orientation, ref_orientation)
+    # apply transformation
+    return nib.orientations.apply_orientation(arr_ras, trans_orient)
```
Comment (andreanne-lemay) [inline ivadomed/image.py:26]: New line missing

File: ivadomed/image.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charleygros) [reply ivadomed/image.py:1]: I've heard it is going to travel towards `loader/utils.py`, which I agree with :+1: !

**Review by charleygros — COMMENTED** [review]
Looks way better this way! Many thanks for this important refactoring @lrouhier

**Review by charleygros — APPROVED** [review]

**Pull request merged**